src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/timeline/BookingTimelineService.kt

```diff
@@ -209,6 +224,153 @@ class BookingTimelineService(
       .groupBy { it.date }
   }
 
+  private fun applyTrancheAdjustmentLogic(
+    calculation: CalculationResult,
+    adjustments: Adjustments,
+    trancheAndCommencement: Pair<SDSEarlyReleaseTranche, LocalDate?>,
+  ): CalculationResult {
+    if (trancheAndCommencement.first !in listOf(
+        SDSEarlyReleaseTranche.TRANCHE_1,
+        SDSEarlyReleaseTranche.TRANCHE_2,
+      )
+    ) {
+      log.info("No adjustments to apply as tranche is ${calculation.sdsEarlyReleaseTranche}")
+      return calculation
+    }
+
+    return adjustReleaseDatesForTranche(calculation, adjustments, trancheAndCommencement.second!!)
+  }
+
+  private fun adjustReleaseDatesForTranche(
+    calculation: CalculationResult,
+    adjustments: Adjustments,
+    commencementDate: LocalDate,
+  ): CalculationResult {
+    val rulesToHandle = listOf(
+      CalculationRule.SDS_EARLY_RELEASE_ADJUSTED_TO_TRANCHE_1_COMMENCEMENT,
+      CalculationRule.SDS_EARLY_RELEASE_ADJUSTED_TO_TRANCHE_2_COMMENCEMENT,
+      CalculationRule.SDS_STANDARD_RELEASE_APPLIES,
+    )
+
+    val releaseDateTypesToAdjust = listOf(
+      ReleaseDateType.HDCED,
+      ReleaseDateType.ERSED,
+      ReleaseDateType.PED,
+    )
+
+    var updatedCalculation = calculation
+
+    for (rule in rulesToHandle) {
+      for (releaseDateType in releaseDateTypesToAdjust) {
+        updatedCalculation = adjustReleaseDateForType(
+          updatedCalculation,
+          adjustments,
+          commencementDate,
+          releaseDateType,
+          rule,
+        )
+      }
+    }
+
+    return updatedCalculation
+  }
+
+  private fun adjustReleaseDateForType(
+    calculation: CalculationResult,
+    adjustments: Adjustments,
+    commencementDate: LocalDate,
+    releaseDateType: ReleaseDateType,
+    rule: CalculationRule,
+  ): CalculationResult {
+    val applicableRules = calculation.breakdownByReleaseDateType[releaseDateType]?.rules
+
+    if (applicableRules?.contains(rule) == true) {
+      log.info("${releaseDateType.name} before adjustment: ${calculation.dates[releaseDateType]}")
+      val crd = calculation.dates[ReleaseDateType.CRD]!!
```

Comment:
Is there always a CRD? Could we get here with an ARD or DTO release types? Maybe one to check with Cherky.

Reply:
Or at least only apply this logic if the CRD is non-null.

Reply:
added a defensive check against the CRD being null... will consult Cherky on wheher CRD or DTO release types should also be considered. Will keep you posted but if any changes are required will do that in the follow up ticket that Cherky is drafting